cuda_core/tests/test_module.py

```diff
@@ -162,3 +174,85 @@ def test_object_code_load_cubin_from_file(get_saxpy_kernel, tmp_path):
     mod = get_saxpy_object_code
     assert mod.handle is not None
+
+
+@skipif_testing_with_compute_sanitizer
+def test_saxpy_arguments(get_saxpy_kernel, cuda12_prerequisite_check):
+    if not cuda12_prerequisite_check:
+        pytest.skip("Test requires CUDA 12")
+    krn, _ = get_saxpy_kernel
+
+    assert krn.num_arguments == 5
+
+    assert "ParamInfo" in str(type(krn).arguments_info.fget.__annotations__)
+    arg_info = krn.arguments_info
+    n_args = len(arg_info)
+    assert n_args == krn.num_arguments
+
+    class ExpectedStruct(ctypes.Structure):
+        _fields_ = [
+            ("a", ctypes.c_float),
+            ("x", ctypes.POINTER(ctypes.c_float)),
+            ("y", ctypes.POINTER(ctypes.c_float)),
+            ("out", ctypes.POINTER(ctypes.c_float)),
+            ("N", ctypes.c_size_t),
+        ]
+
+    offsets = [p.offset for p in arg_info]
+    sizes = [p.size for p in arg_info]
+    members = [getattr(ExpectedStruct, name) for name, _ in ExpectedStruct._fields_]
+    expected_offsets = tuple(m.offset for m in members)
+    assert all(actual == expected for actual, expected in zip(offsets, expected_offsets))
+    expected_sizes = tuple(m.size for m in members)
+    assert all(actual == expected for actual, expected in zip(sizes, expected_sizes))
+
+
+@skipif_testing_with_compute_sanitizer
+@pytest.mark.parametrize("nargs", [0, 1, 2, 3, 16])
+@pytest.mark.parametrize("c_type_name,c_type", [("int", ctypes.c_int), ("short", ctypes.c_short)], ids=["int", "short"])
+def test_num_arguments(init_cuda, nargs, c_type_name, c_type, cuda12_prerequisite_check):
+    if not cuda12_prerequisite_check:
+        pytest.skip("Test requires CUDA 12")
+    args_str = ", ".join([f"{c_type_name} p_{i}" for i in range(nargs)])
+    src = f"__global__ void foo{nargs}({args_str}) {{ }}"
+    prog = Program(src, code_type="c++")
+    mod = prog.compile(
+        "cubin",
+        name_expressions=(f"foo{nargs}",),
+    )
+    krn = mod.get_kernel(f"foo{nargs}")
+    assert krn.num_arguments == nargs
+
+    class ExpectedStruct(ctypes.Structure):
+        _fields_ = [(f"arg_{i}", c_type) for i in range(nargs)]
+
+    members = tuple(getattr(ExpectedStruct, f"arg_{i}") for i in range(nargs))
+
+    arg_info = krn.arguments_info
+    assert all([actual.offset == expected.offset for actual, expected in zip(arg_info, members)])
+    assert all([actual.size == expected.size for actual, expected in zip(arg_info, members)])
+
+
+@skipif_testing_with_compute_sanitizer
+def test_num_args_error_handling(deinit_context_function, cuda12_prerequisite_check):
+    if not cuda12_prerequisite_check:
+        pytest.skip("Test requires CUDA 12")
+    src = "__global__ void foo(int a) { }"
+    prog = Program(src, code_type="c++")
+    mod = prog.compile(
+        "cubin",
+        name_expressions=("foo",),
+    )
+    krn = mod.get_kernel("foo")
+    # Unset current context using function from conftest
+    while True:
```

Comment:
Question: Could this `while` loop be moved into `deinit_context_function`?

Suggestion (outline only):

```
    for _ in range(100):
        ... code you have here ...
    else:
        raise RuntimeError("Unexpected number of loop iterations.")
```

This is to prevent a stuck CI in case there is a bug elsewhere. (Been there several times.)